### PR TITLE
Merge upstream 3.2.0

### PR DESCRIPTION
**Pull request by manicmaniac**
## Overview

- Merged upstream/3.2.0 into master at https://github.com/quipper/MessageKit/pull/22/commits/3fde1580a1632d42ab74233309c48d7f0fabf0c0
- Resolved a number of conflicts at https://github.com/quipper/MessageKit/pull/22/commits/d79d9397c0d05346df20e1c2235e4a440ec0f5fb and https://github.com/quipper/MessageKit/pull/22/commits/ab772c46997001780e2548c7909da0a22335a9d8
- Fixed unit test at https://github.com/quipper/MessageKit/pull/22/commits/c08c3dc7782a343a8464e310754219f13e5c49af
- Resolved CI problems at https://github.com/quipper/MessageKit/pull/22/commits/86bccf00296e658e949243b27b02adf664059289 and the later commits

You don't need to see other commits.

## Known issues

Example app cannnot be built as reported in https://github.com/MessageKit/MessageKit/pull/1441

To workaround this, cherry-pick the PR.

```sh
git cherry-pick 94fd0709c1b901dd0e6586670d879c1517d2442f
```

## Screenshot

image

### CODE REVIEW

File: Example/Sources/View Controllers/LaunchViewController.swift [hunk 1]
```diff
@@ -33,11 +33,7 @@ final internal class LaunchViewController: UITableViewController {
         return .lightContent
     }
 
-<<<<<<< ours
-    let cells = ["Announcement Example", "Basic Example", "Advanced Example", "Autocomplete Example", "Embedded Example", "Settings", "Source Code", "Contributors"]
-=======
-    let cells = ["Basic Example", "Advanced Example", "Autocomplete Example", "Embedded Example", "Subview Example", "SwiftUI Example", "Settings", "Source Code", "Contributors"]
->>>>>>> theirs
+    let cells = ["Announcement Example", "Basic Example", "Advanced Example", "Autocomplete Example", "Embedded Example", "Subview Example", "SwiftUI Example", "Settings", "Source Code", "Contributors"]
```
Comment (manicmaniac) [inline Example/Sources/View Controllers/LaunchViewController.swift:36]: `Announcement` feature was added by @m-sugawara.

File: Sources/Models/MessageStyle.swift [hunk 2]
```diff
@@ -26,91 +26,8 @@ import UIKit
 
 public enum MessageStyle {
     case none
-<<<<<<< ours
     case leftBubble
     case rightBubble
     case announcement
     case warning(color: UIColor)
```
Comment (manicmaniac) [inline Sources/Models/MessageStyle.swift]: We heavily customize `MessageStyle`. Upstream changes are simply ignored.

File: Tests/ViewsTests/AvatarViewTests.swift [hunk 3]
```diff
@@ -50,7 +50,7 @@ class AvatarViewTests: XCTestCase {
     func testWithImage() {
         let avatar = Avatar(image: nil)
         avatarView.set(avatar: avatar)
-        XCTAssertEqual(avatar.initials, "?")
+        XCTAssertEqual(avatar.initials, "")
```
Comment (manicmaniac) [inline Tests/ViewsTests/AvatarViewTests.swift:53]: We changed the default value at https://github.com/quipper/MessageKit/commit/de95cd4e85f91566b9cbf0e7b9d6e9f27fb443f0#diff-8272c0fbb141fe6938bead67a2365ce451c7ecb42e0107acc6443ce23e7d2581

File: GitHubActions/build.sh [hunk 4]
```diff
@@ -35,14 +35,14 @@ MODE="$1"
 
 if [ "$MODE" = "tests" -o "$MODE" = "all" ]; then
   echo "Running MessageKit tests."
-  carthage bootstrap --platform ios
+  carthage checkout && ./GitHubActions/install-dependencies.sh
```
Comment (manicmaniac) [inline GitHubActions/build.sh]: Back then, Carthage did not support xcframework.
Nowadays we need to use xcframework to build on both Intel Mac and Apple Silicon like `carthage bootstrap --platform ios --use-xcframeworks`.

However, to use xcframeworks need to vastly change project settings, which leads conflicts in future update.

So I added `./GitHubActions/install-dependencies.sh` script to build frameworks using `xcodebuild` directly.

File: GitHubActions/install-dependencies.sh [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+        -project "$project" \
+        -scheme "$scheme" \
+        -configuration Debug \
+        -derivedDataPath "$derived_data_dir" \
+        -sdk iphonesimulator \
+        -quiet \
+        ONLY_ACTIVE_ARCH=YES \
+        CODE_SIGNING_REQUIRED=NO \
+        CODE_SIGN_IDENTITY= \
+        CARTHAGE=YES
```
Comment (manicmaniac) [inline GitHubActions/install-dependencies.sh:34]: This is almost same as Carthage does but this script builds framework for active architecture only.

File: .github/workflows/ci_pr_example.yml [hunk 6]
```diff
@@ -5,7 +5,7 @@ on: pull_request
 jobs:
   tests:
     name: Build Example app
-    runs-on: macOS-latest
+    runs-on: macOS-11
```
Comment (manicmaniac) [inline .github/workflows/ci_pr_example.yml:8]: The current `macOS-latest` is equivalent to `macOS-12`.
At that time, `macOS-11` should be the latest.

upstream/3.2.0 dependes on Ruby 2.7 which is not installed in `macOS-12` but `macOS-11`.

File: GitHubActions/build.sh [hunk 7]
```diff
@@ -53,7 +53,7 @@ if [ "$MODE" = "example" -o "$MODE" = "all" ]; then
   gem install bundler
   bundle check || bundle install
   bundle exec pod install
-  set -o pipefail && xcodebuild build analyze -workspace ChatExample.xcworkspace -scheme ChatExample -destination "platform=iOS Simulator,name=iPhone 11 Pro" CODE_SIGNING_REQUIRED=NO | xcpretty -c
+  set -o pipefail && xcodebuild build analyze -workspace ChatExample.xcworkspace -scheme ChatExample -destination "platform=iOS Simulator,name=iPhone 11 Pro" CODE_SIGNING_REQUIRED=NO OTHER_CFLAGS="-Xclang -fcompatibility-qualified-id-block-type-checking" | xcpretty -c
```
Comment (manicmaniac) [inline GitHubActions/build.sh:56]: One of the dependencies `PINRemoteImage` is not compatible with Xcode 12.
The compiler complains that there's incompatible block pointers.

So I added workaround to disable this warning by passing extra flags.

File: Example/Podfile [hunk 8]
```diff
@@ -16,3 +16,18 @@ target 'ChatExample' do
     inherit! :search_paths
   end
 end
+
+# Workaround for https://github.com/pinterest/PINRemoteImage/issues/566
+post_install do |installer|
+  installer.pods_project.targets.each do |target|
+    next unless %w[PINCache PINRemoteImage].include?(target.name)
+
+    target.build_configurations.each do |config|
```
Comment (manicmaniac) [inline Example/Podfile:25]: One of the dependencies `PINRemoteImage` (and its subspec PINCache) is not compatible with Xcode 12.
The compiler complains that there's incompatible block pointers.

So I added workaround to disable this warning by passing extra flags.

File: Example/Podfile [hunk 9]
```diff
@@ -16,3 +16,18 @@ target 'ChatExample' do
...
+      config.build_settings['OTHER_CFLAGS'] = [
+        '$(inherited)',
+        '-Xclang',
+        '-fcompatibility-qualified-id-block-type-checking'
+      ].join(' ')
+    end
+  end
+end
```
Comment (manicmaniac) [inline Example/Podfile]: One of the dependencies `PINRemoteImage` (and its subspec PINCache) is not compatible with Xcode 12.
The compiler complains that there's incompatible block pointers.

So I added workaround to disable this warning by passing extra flags.

**Review by k-kohey — APPROVED** [review]
LGTM

**Review by elliekwon — APPROVED** [review]
Overviewに書かれてるcommitで疑問はありません！LGTM🚀

**Pull request merged**